internal/util/fencing.go

```diff
@@ -0,0 +1,142 @@
+package util
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/csi-addons/spec/lib/go/fence"
+	"github.com/golang/protobuf/ptypes/timestamp"
+)
+
+type NetworkFence struct {
+	Cidr      string
+	Monitors  string
+	CreatedAt *timestamp.Timestamp
+}
+
+// addCephBlocklist adds an IP to ceph osd blocklist.
+func (nwFence *NetworkFence) addCephBlocklist(ctx context.Context, ip string, cr *Credentials) error {
+	arg := []string{
+		"--id", cr.ID,
+		"--keyfile=" + cr.KeyFile,
+		"-m", nwFence.Monitors,
+	}
+	cmd := []string{"osd", "blocklist", "add"}
+	cmd = append(cmd, arg...)
+
+	_, stderr, err := ExecCommand(ctx, "ceph", cmd...)
+	if err != nil {
+		return fmt.Errorf("failed to blocklist IP %s: %v", ip, stderr)
+	}
+
+	return nil
+}
+
+// AddNetworkFence blocks access for all the IPs in the IP range mentioned via the CIDR block
+// using a network fence.
+func AddNetworkFence(ctx context.Context, nwFence *NetworkFence, cr *Credentials) error {
+	// fetch the list of IPs from a CIDR block
+	hosts, err := getIPRange(nwFence.Cidr)
+	if err != nil {
+		return fmt.Errorf("failed to convert cidr block to corresponding IP range")
+	}
+	// add ceph blocklist for each IP in the range mentioned by the CIDR
+	for _, host := range hosts {
+		err = nwFence.addCephBlocklist(ctx, host, cr)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getIpRange returns a list of IPs from the IP range
+// corresponding to a CIDR block.
+func getIPRange(netw string) ([]string, error) {
+	// convert string to IPNet struct
+	var infinity uint32 = 0xffffffff
+	_, ipv4Net, err := net.ParseCIDR(netw)
+	if err != nil {
+		return nil, err
+	}
+	// convert IPNet struct mask and address to uint32
+	mask := binary.BigEndian.Uint32(ipv4Net.Mask)
+	// start IP address
+	start := binary.BigEndian.Uint32(ipv4Net.IP)
```

Comment:
can you do something like this https://go.dev/play/p/m8TNTtygK0 to get range in CIRD